# Chapter 6: Direct Methods for Solving Linear Systems

$$\text{葛雨辰 201800150053}$$

## 1 Introduction

In the homework, I use Jupyter notebook ( Python Kernel ) to run the codes and show the results. 

## 2 Gaussian Elimination Method (6.1/4c,4d)

We will apply Gaussian Elimination Method with backward substitution for problems 6.1/4c,4d. We run the program as follows.

In [156]:
import numpy as np
C_old=np.array([[1,1/2,1/3,1/4,1/6],[1/2,1/3,1/4,1/5,1/7],[1/3,1/4,1/5,1/6,1/8],[1/4,1/5,1/6,1/7,1/9]])
D_old=np.array([[2,1,-1,1,-3,7],[1,0,2,-1,1,2],[0,-2,-1,1,-1,-5],[3,1,-4,0,5,6],[1,-1,-1,-1,1,3]])
C=np.c_[np.random.randn(4,1),C_old]
C=np.r_[np.random.randn(1,6),C]
D=np.c_[np.random.randn(5,1),D_old]
D=np.r_[np.random.randn(1,7),D]

def Gaussian(n,A):
    x=np.random.randn(n+1,1)
    for i in range(1,n):
        # find the smallest p
        temp=list(A[i:n+1,i])
        if temp==0:
            print('no unique solution exists')
            return
        else:
            temporary=min(filter(lambda x: x!=0, temp))
            p=temp.index(temporary)+i
            
        if p!=i:
            A[[p, i],:] = A[[i, p],:]
        for j in range(i+1,n+1):
            h=A[j,i]/A[i,i]
            tempal=A[j,:]-h*A[i,:]
            A[j,:]=tempal
    if A[n,n]==0:
        print('no unique solution exists')
        return
    x[n]=A[n,n+1]/A[n,n]
    for i in range(1,n):
        temp=0
        for j in range(n-i+1,n+1):
            temp=temp+A[n-i,j]*x[j]
        x[n-i]=(A[n-i,n+1]-temp)/A[n-i,n-i]
    return (x[1:n+1]).T
        
Gaussian(4,C), Gaussian(5,D)

(array([[-0.03174603,  0.5952381 , -2.38095238,  2.77777778]]),
 array([[ 1.91812865,  1.96491228, -0.98830409, -3.19298246, -1.13450292]]))

Therefore, we solve the problem 6.1/4c and show the results in the table below.

| Problem |  6.1/4c |
|  ----  | ----  | 
| $x_1$ | -0.03174603 |
| $x_2$  | 0.5952381 |
| $x_3$  |  -2.38095238 |
| $x_4$  | 2.77777778 |


Moreover, we also show the results of problem 6.1/4c in the table below.

| Problem | 6.1/4d |
|  ----  | ----  |  
| $x_1$ |1.91812865 | 
| $x_2$  | 1.96491228 | 
| $x_3$  |  -0.98830409 |
| $x_4$  | -3.19298246 |
| $x_5$  | -1.13450292 |

We see that the Gaussian Elimination Method with backward substitution outputs the correct results for problems 6.1/4c,4d by checking as below which is the same as the right-hand sides of the linear systems.

In [157]:
np.dot(C_old[:,0:4], Gaussian(4,C).T), np.dot(D_old[:,0:5], Gaussian(5,D).T) #check

(array([[0.16666667],
        [0.14285714],
        [0.125     ],
        [0.11111111]]),
 array([[ 7.],
        [ 2.],
        [-5.],
        [ 6.],
        [ 3.]]))

## 3 Gaussian Partial Pivoting Method (6.2/5e,6,8,10)

Since Python can not do the operation of rounding, we are not in a position to do 6.2/6. 

But We will apply Gaussian Partial Pivoting Method ( for problem 6.2/8 ) and Gaussian Scaled Partial Pivoting Method ( for problem 6.2/10 ). And we apply Maple for problem 6.2/5e. We run the program as follows.

In [159]:
import numpy as np
import scipy
E_old=np.array([[1.19,2.11,-100,1,1.12],[14.2,-0.122,12.2,-1,3.44],[0,100,-99.9,1,2.15],[15.3,0.11,-13.1,-1,4.16]])
E=np.c_[np.random.randn(4,1),E_old]
E=np.r_[np.random.randn(1,6),E]

np.set_printoptions(precision = 8,suppress = True)
# Gaussian Elimination with Partial Pivoting
def GaussianPartialPivoting(n,A):
    NROW=list(range(0,n+1))
    x=list(range(0,n+1))
    for i in range(1,n+1):
        NROW[i]=i
    for i in range(1,n):
        # actually to realize temp=list(A[NROW(i:n+1),i])
        temp=list(range(i,n+1))
        for j in range(i,n+1):
            temp[j-i]=A[NROW[j],i]  
        temp=np.array(temp)
        temporary=max(abs(temp))
        p=list(abs(temp)).index(temporary)+i
        if A[NROW[p],i]==0:
            print('no unique solutions exists')
            return
        if NROW[i]!=NROW[p]:
            NCOPY=NROW[i]
            NROW[i]=NROW[p]
            NROW[p]=NCOPY
        for j in range(i+1,n+1):
            temporary=A[NROW[j],i]/A[NROW[i],i]
            A[NROW[j],:]=A[NROW[j],:]-temporary*A[NROW[i],:]
    if A[NROW[n],n]==0:
        print('no unique solution exists')
        return
    x[n]=A[NROW[n],n+1]/A[NROW[n],n]
    currlist=list(range(1,n))
    currlist.reverse()
    for i in currlist:
        temp=0
        for j in range(i+1,n+1):
            temp=temp+A[NROW[i],j]*x[j]
        x[i]=(A[NROW[i],n+1]-temp)/A[NROW[i],i]
    print('Result of GaussianPartialPivoting method (6.2/8) is',(np.array(x[1:n+1])).T)
    return (np.array(x[1:n+1])).T

# Gaussian Elimination with Scaled Partial Pivoting
def GaussianScaledPartialPivoting(n,A):
    NROW=list(range(0,n+1))
    x=list(range(0,n+1))
    s=list(range(0,n+1))
    for i in range(1,n+1):
        s[i]=max(abs(A[i,:]))
    if s==0:
        print('no unique solutions exists')
        return
    for i in range(1,n+1):
        NROW[i]=i
    for i in range(1,n):
        # actually to realize temp=list(A[NROW(i:n+1),i])
        temp=list(range(i,n+1))
        for j in range(i,n+1):
            temp[j-i]=A[NROW[j],i]/s[NROW[j]] 
        temp=np.array(temp)
        temporary=max(abs(temp))
        p=list(abs(temp)).index(temporary)+i
        if A[NROW[p],i]==0:
            print('no unique solutions exists')
            return
        if NROW[i]!=NROW[p]:
            NCOPY=NROW[i]
            NROW[i]=NROW[p]
            NROW[p]=NCOPY
        for j in range(i+1,n+1):
            temporary=A[NROW[j],i]/A[NROW[i],i]
            A[NROW[j],:]=A[NROW[j],:]-temporary*A[NROW[i],:]
    if A[NROW[n],n]==0:
        print('no unique solution exists')
        return
    x[n]=A[NROW[n],n+1]/A[NROW[n],n]
    currlist=list(range(1,n))
    currlist.reverse()
    for i in currlist:
        temp=0
        for j in range(i+1,n+1):
            temp=temp+A[NROW[i],j]*x[j]
        x[i]=(A[NROW[i],n+1]-temp)/A[NROW[i],i]
    print('Result of GaussianScaledPartialPivoting method (6.2/10) is',(np.array(x[1:n+1])).T)
    return (np.array(x[1:n+1])).T

GaussianPartialPivoting(4,E)
GaussianScaledPartialPivoting(4,E)

Result of GaussianPartialPivoting method (6.2/8) is [ 0.1768253   0.01269269 -0.02065405 -1.1826087 ]
Result of GaussianScaledPartialPivoting method (6.2/10) is [ 0.1768253   0.01269269 -0.02065405 -1.1826087 ]


array([ 0.1768253 ,  0.01269269, -0.02065405, -1.1826087 ])

Here's the table for the results:

|  Method  | Gaussian Method using three digit rounding (6.2/5e) | Gaussian Partial Pivoting Method (6.2/8)  | Gaussian Scaled Partial Pivoting Method (6.2/10)  | 
|  ----  | ----  | ----  |  ----  | 
| $x_1$ | 0.206  | 0.1768253 | 0.1768253 | 
| $x_2$  | 0.0154 |  0.01269269 | 0.01269269 | 
| $x_3$  | -0.0156  | -0.02065405 | -0.02065405 |
| $x_4$  | -0.716  | -1.1826087 | -1.1826087 | 

We see that the Gaussian Partial Pivoting Method (6.2/8) and the  Gaussian Scaled Partial Pivoting Method (6.2/10) output the correct values. While the usual Gaussian method using three digit rounding gives the wrong result since the roundoff error is too big.

## 4 LU decomposition (6.5/3d)

We will apply LU decomposition algorithm ( for problem 6.5/3d ). We run the program as follows.

In [160]:
# LU decomposition
import numpy as np
def LU_Factorization(A):
    n = A.shape[0]
    L = np.zeros([n, n])
    U = np.zeros([n, n])
    for i in range(n):
        L[i][i] = 1
        if i == 0:
            U[0][0] = A[0][0]
            for j in range(1, n):
                U[0][j] = A[0][j]
                L[j][0] = A[j][0] / U[0][0]
        else:
            for j in range(i, n):
                temp = 0
                for k in range(0, i):
                    temp = temp + L[i][k] * U[k][j]
                U[i][j] = A[i][j] - temp
            for j in range(i + 1, n):
                temp = 0
                for k in range(0, i):
                    temp = temp + L[j][k] * U[k][i]
                L[j][i] = (A[j][i] - temp) / U[i][i]
    return L, U
def main():
    D = np.array([[2.1756, 4.0231, -2.1732, 5.1967], [-4.0231,6,0,1.1973], [-1,-5.2107,1.1111,0],[6.0235,7,0,-4.1561]])
    L, U = LU_Factorization(D)
    print('L matrix of 6.5/3d is \n',L, '\nU matrix of 6.5/3d is \n', U)
main()

L matrix of 6.5/3d is 
 [[ 1.          0.          0.          0.        ]
 [-1.84919103  1.          0.          0.        ]
 [-0.45964332 -0.25012194  1.          0.        ]
 [ 2.76866152 -0.30794361 -5.35228302  1.        ]] 
U matrix of 6.5/3d is 
 [[ 2.1756      4.0231     -2.1732      5.1967    ]
 [ 0.         13.43948042 -4.01866194 10.80699101]
 [ 0.          0.         -0.89295239  5.09169403]
 [ 0.          0.          0.         12.03612803]]


We see that the LU decomposition algorithm ( for problem 6.5/3d ) outputs the correct values with an easy check as before.

## 5 LDL' and Choleski's decomposition (6.6/2d,3)

Finally, we are in a position to apply $LDL^{T}$ and $Choleski$'s decomposition to the questions 6.6/2d,3. We run the program as follows. 

In [161]:
import numpy as np
# LDLT decomposition
def LDLT(amatrix):
    if len(np.shape(amatrix)) != 2 or np.shape(amatrix)[0]!=np.shape(amatrix)[1]:
        print("error shape")
        return 
    for i in range(np.shape(amatrix)[0]):
        for j in range(np.shape(amatrix)[1]):
            if amatrix[i][j] != amatrix[j][i]:
                print("The input matrix should be symmetric")
                return 
        
    n = np.shape(amatrix)[0] #dimension of matrix
    l = np.eye(n)
    d = np.zeros((n,n))
    for k in range(n):
        if k == 0:
            d[k][k] = amatrix[k][k]
            if d[k][k] == 0:
                print('error matrix type with 0 sequential principal minor determinant')
                return        
            for m in range(n):
                l[m][k] = amatrix[m][k]/d[k][k]
        else:
            temp_sum1 = 0
            for m in range(k):
                temp_sum1 += amatrix[k][m]*l[k][m]
            d[k][k] = amatrix[k][k] - temp_sum1
            for j in range(k+1,n):
                temp_sum2 = 0
                for m in range(k):
                    temp_sum2 += amatrix[j][m]*l[k][m]
                amatrix[j][k] = amatrix[j][k] - temp_sum2
                l[j][k] = amatrix[j][k]/d[k][k]
    print("L of LDL' decomposition in 6.6/2d is")
    print(l)
    print("D of LDL' decomposition in 6.6/2d is")
    print(d)
    return l,d

# Cholesky Decomposition
def Cholesky(matrix):
    w = matrix.shape[0]
    G = np.zeros((w,w))
    for i in range(w):
        G[i,i] = (matrix[i,i] - np.dot(G[i,:i],G[i,:i].T))**0.5
        for j in range(i+1,w):
            G[j,i] = (matrix[j,i] - np.dot(G[j,:i],G[i,:i].T))/G[i,i]
    print("L of Cholesky decomposition in 6.6/3 is")
    print(G)
    return G
   
# decomposition of 6.6/2d
A=np.array([[6,2,1,-1],[2,4,1,0],[1,1,4,-1],[-1,0,-1,3]],dtype=float)
L,D = LDLT(A)

AA=np.array([[6,2,1,-1],[2,4,1,0],[1,1,4,-1],[-1,0,-1,3]],dtype=float)
LL = Cholesky(AA)

L of LDL' decomposition in 6.6/2d is
[[ 1.          0.          0.          0.        ]
 [ 0.33333333  1.          0.          0.        ]
 [ 0.16666667  0.2         1.          0.        ]
 [-0.16666667  0.1        -0.24324324  1.        ]]
D of LDL' decomposition in 6.6/2d is
[[6.         0.         0.         0.        ]
 [0.         3.33333333 0.         0.        ]
 [0.         0.         3.7        0.        ]
 [0.         0.         0.         2.58108108]]
L of Cholesky decomposition in 6.6/3 is
[[ 2.44948974  0.          0.          0.        ]
 [ 0.81649658  1.82574186  0.          0.        ]
 [ 0.40824829  0.36514837  1.92353841  0.        ]
 [-0.40824829  0.18257419 -0.46788772  1.60657433]]


Finally we see that the $LDL^{T}$ and $Choleski$'s decomposition algorithm output the correct values with an easy check.

We check as follows and find that the decomposition above gives exactly the correct result.

In [176]:
np.matmul(np.matmul(L,D),L.T)-A
np.matmul(LL,LL.T)-AA

array([[-0.,  0.,  0.,  0.],
       [ 0., -0., -0.,  0.],
       [ 0., -0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.]])

## 6 Appendix of Codes

### 6.1  Gaussian Elimination Method 

In [148]:
import numpy as np
C_old=np.array([[1,1/2,1/3,1/4,1/6],[1/2,1/3,1/4,1/5,1/7],[1/3,1/4,1/5,1/6,1/8],[1/4,1/5,1/6,1/7,1/8]])
D_old=np.array([[2,1,-1,1,-3,7],[1,0,2,-1,1,2],[1,-2,-1,1,-1,-5],[3,1,-4,0,5,6],[1,-1,-1,-1,1,3]])
C=np.c_[np.random.randn(4,1),C_old]
C=np.r_[np.random.randn(1,6),C]
D=np.c_[np.random.randn(5,1),D_old]
D=np.r_[np.random.randn(1,7),D]

def Gaussian(n,A):
    x=np.random.randn(n+1,1)
    for i in range(1,n):
        # find the smallest p
        temp=list(A[i:n+1,i])
        if temp==0:
            print('no unique solution exists')
            return
        else:
            temporary=min(filter(lambda x: x!=0, temp))
            p=temp.index(temporary)+i
            
        if p!=i:
            A[[p, i],:] = A[[i, p],:]
        for j in range(i+1,n+1):
            h=A[j,i]/A[i,i]
            tempal=A[j,:]-h*A[i,:]
            A[j,:]=tempal
    if A[n,n]==0:
        print('no unique solution exists')
        return
    x[n]=A[n,n+1]/A[n,n]
    for i in range(1,n):
        temp=0
        for j in range(n-i+1,n+1):
            temp=temp+A[n-i,j]*x[j]
        x[n-i]=(A[n-i,n+1]-temp)/A[n-i,n-i]
    return (x[1:n+1]).T

### 6.2 Gaussian Partial Pivoting Method 

In [152]:
import numpy as np
import scipy
E_old=np.array([[1.19,2.11,-100,1,1.12],[14.2,-0.122,12.2,-1,3.44],[0,100,-99.9,1,2.15],[15.3,0.11,-13.1,-1,4.16]])
E=np.c_[np.random.randn(4,1),E_old]
E=np.r_[np.random.randn(1,6),E]

#numpy.set_printoptions(precision = 8,suppress = True)
# Gaussian Elimination with Partial Pivoting
def GaussianPartialPivoting(n,A):
    NROW=list(range(0,n+1))
    x=list(range(0,n+1))
    for i in range(1,n+1):
        NROW[i]=i
    for i in range(1,n):
        # actually to realize temp=list(A[NROW(i:n+1),i])
        temp=list(range(i,n+1))
        for j in range(i,n+1):
            temp[j-i]=A[NROW[j],i]  
        temp=np.array(temp)
        temporary=max(abs(temp))
        p=list(abs(temp)).index(temporary)+i
        if A[NROW[p],i]==0:
            print('no unique solutions exists')
            return
        if NROW[i]!=NROW[p]:
            NCOPY=NROW[i]
            NROW[i]=NROW[p]
            NROW[p]=NCOPY
        for j in range(i+1,n+1):
            temporary=A[NROW[j],i]/A[NROW[i],i]
            A[NROW[j],:]=A[NROW[j],:]-temporary*A[NROW[i],:]
    if A[NROW[n],n]==0:
        print('no unique solution exists')
        return
    x[n]=A[NROW[n],n+1]/A[NROW[n],n]
    currlist=list(range(1,n))
    currlist.reverse()
    for i in currlist:
        temp=0
        for j in range(i+1,n+1):
            temp=temp+A[NROW[i],j]*x[j]
        x[i]=(A[NROW[i],n+1]-temp)/A[NROW[i],i]
    print('Result of GaussianPartialPivoting method (6.2/8) is',(np.array(x[1:n+1])).T)
    return (np.array(x[1:n+1])).T

# Gaussian Elimination with Scaled Partial Pivoting
def GaussianScaledPartialPivoting(n,A):
    NROW=list(range(0,n+1))
    x=list(range(0,n+1))
    s=list(range(0,n+1))
    for i in range(1,n+1):
        s[i]=max(abs(A[i,:]))
    if s==0:
        print('no unique solutions exists')
        return
    for i in range(1,n+1):
        NROW[i]=i
    for i in range(1,n):
        # actually to realize temp=list(A[NROW(i:n+1),i])
        temp=list(range(i,n+1))
        for j in range(i,n+1):
            temp[j-i]=A[NROW[j],i]/s[NROW[j]] 
        temp=np.array(temp)
        temporary=max(abs(temp))
        p=list(abs(temp)).index(temporary)+i
        if A[NROW[p],i]==0:
            print('no unique solutions exists')
            return
        if NROW[i]!=NROW[p]:
            NCOPY=NROW[i]
            NROW[i]=NROW[p]
            NROW[p]=NCOPY
        for j in range(i+1,n+1):
            temporary=A[NROW[j],i]/A[NROW[i],i]
            A[NROW[j],:]=A[NROW[j],:]-temporary*A[NROW[i],:]
    if A[NROW[n],n]==0:
        print('no unique solution exists')
        return
    x[n]=A[NROW[n],n+1]/A[NROW[n],n]
    currlist=list(range(1,n))
    currlist.reverse()
    for i in currlist:
        temp=0
        for j in range(i+1,n+1):
            temp=temp+A[NROW[i],j]*x[j]
        x[i]=(A[NROW[i],n+1]-temp)/A[NROW[i],i]
    print('Result of GaussianScaledPartialPivoting method (6.2/10) is',(np.array(x[1:n+1])).T)
    return (np.array(x[1:n+1])).T

### 6.3 LU decomposition

In [153]:
# LU decomposition
import numpy as np
def LU_Factorization(A):
    n = A.shape[0]
    L = np.zeros([n, n])
    U = np.zeros([n, n])
    for i in range(n):
        L[i][i] = 1
        if i == 0:
            U[0][0] = A[0][0]
            for j in range(1, n):
                U[0][j] = A[0][j]
                L[j][0] = A[j][0] / U[0][0]
        else:
            for j in range(i, n):
                temp = 0
                for k in range(0, i):
                    temp = temp + L[i][k] * U[k][j]
                U[i][j] = A[i][j] - temp
            for j in range(i + 1, n):
                temp = 0
                for k in range(0, i):
                    temp = temp + L[j][k] * U[k][i]
                L[j][i] = (A[j][i] - temp) / U[i][i]
    return L, U
#def main():
#    D = np.array([[2.1756, 4.0231, -2.1732, 5.1967], [-4.0231,6,0,1.1973], [-1,-5.2107,1.1111,0],[6.0235,7,0,-4.1561]])
#    L, U = LU_Factorization(D)
#    print('L matrix of 6.5/3d is \n',L, '\nU matrix of 6.5/3d is \n', U)
#main()

### 6.4 LDLT decomposition and Cholesky decomposition

In [178]:
import numpy as np
# LDLT decomposition
def LDLT(amatrix):
    if len(np.shape(amatrix)) != 2 or np.shape(amatrix)[0]!=np.shape(amatrix)[1]:
        print("error shape")
        return 
    for i in range(np.shape(amatrix)[0]):
        for j in range(np.shape(amatrix)[1]):
            if amatrix[i][j] != amatrix[j][i]:
                print("The input matrix should be symmetric")
                return 
        
    n = np.shape(amatrix)[0] #dimension of matrix
    l = np.eye(n)
    d = np.zeros((n,n))
    for k in range(n):
        if k == 0:
            d[k][k] = amatrix[k][k]
            if d[k][k] == 0:
                print('error matrix type with 0 sequential principal minor determinant')
                return        
            for m in range(n):
                l[m][k] = amatrix[m][k]/d[k][k]
        else:
            temp_sum1 = 0
            for m in range(k):
                temp_sum1 += amatrix[k][m]*l[k][m]
            d[k][k] = amatrix[k][k] - temp_sum1
            for j in range(k+1,n):
                temp_sum2 = 0
                for m in range(k):
                    temp_sum2 += amatrix[j][m]*l[k][m]
                amatrix[j][k] = amatrix[j][k] - temp_sum2
                l[j][k] = amatrix[j][k]/d[k][k]
    print("L of LDL' decomposition in 6.6/2d is")
    print(l)
    print("D of LDL' decomposition in 6.6/2d is")
    print(d)
    return l,d

# Cholesky Decomposition
def Cholesky(matrix):
    w = matrix.shape[0]
    G = np.zeros((w,w))
    for i in range(w):
        G[i,i] = (matrix[i,i] - np.dot(G[i,:i],G[i,:i].T))**0.5
        for j in range(i+1,w):
            G[j,i] = (matrix[j,i] - np.dot(G[j,:i],G[i,:i].T))/G[i,i]
    print("L of Cholesky decomposition in 6.6/3 is")
    print(G)
    return G

# decomposition of 6.6/2d
#A=np.array([[6,2,1,-1],[2,4,1,0],[1,1,4,-1],[-1,0,-1,3]],dtype=float)
#L,D = LDLT(A)

#AA=np.array([[6,2,1,-1],[2,4,1,0],[1,1,4,-1],[-1,0,-1,3]],dtype=float)
#LL = Cholesky(AA)